src/piton/transforms/__init__.py

```diff
@@ -0,0 +1,123 @@
+
+from piton import Patient
+from piton.extractors.omop import OMOP_BIRTH
+
+
+def move_pre_birth(patient: Patient) -> Optional[Patient]:
+    """Move all events to after the birth of a patient."""
+    birth_date = None
+    for event in patient.events:
+        if event.code == OMOP_BIRTH:
+            birth_date = event.start
+
+    if birth_date is None:
+        return None
+
+    new_events = []
+    for event in patient.events:
+        new_start = None
+        new_end = None
+        if event.start < birth_date:
+            new_start = birth_date
+        if event.end and event.end < birth_date:
+            new_end = birth_date
+
+        if new_start or new_end:
+            new_events.append(
+                dataclasses.replace(
+                    event,
+                    start=new_start or event.start,
+                    end=new_end or event.end,
+                )
+            )
+        else:
+            new_events.append(event)
+
+    return Patient(patient_id=patient.patient_id, events=new_events)
+
+
+def remove_short_patients(
+    patient: Patient, min_num_dates: int = 3
+) -> Optional[Patient]:
+    """Remove patients with too few timepoints."""
+    if (
+        len(set(event.start.date() for event in patient.events))
+        <= min_num_dates
+    ):
+        return None
+    else:
+        return patient
+
+
+def remove_nones(patient: Patient) -> Patient:
+    """Remove duplicate codes w/in same day if duplicate code has None value.
+
+    There is no point having a NONE value in a timeline when we have an actual value within the same day.
+
+    This removes those unnecessary NONE values.
+    """
+    has_value: Set[Tuple[int, datetime.date]] = set()
+
+    for event in patient.events:
+        if event.value is not None:
+            has_value.add((event.code, event.start.date()))
+
+    new_events = []
+    for event in patient.events:
+        if (
+            event.value is None
+            and (event.code, event.start.date()) in has_value
+        ):
+            continue
+        new_events.append(event)
+
+    return Patient(patient.patient_id, new_events)
+
+
+def move_to_day_end(patient: Patient) -> Patient:
+    """We assume that everything coded at midnight should actually be moved to the end of the day."""
+    new_events = []
+    for event in patient.events:
+        if (
+            event.start.hour == 0
+            and event.start.minute == 0
+            and event.start.second == 0
+            and event.code != OMOP_BIRTH
+        ):
+            new_time = (
+                event.start
+                + datetime.timedelta(days=1)
+                - datetime.timedelta(seconds=1)
+            )
+            new_events.append(dataclasses.replace(event, start=new_time))
+        else:
+            new_events.append(event)
+
+    new_events.sort(key=lambda a: (a.start, a.code))
+
+    return Patient(patient.patient_id, new_events)
+
+
+def delta_encode(patient: Patient) -> Patient:
```

Comment:
I may be completely misunderstanding it, but I was thinking that events can be anything like lab measurements or vital signs, which are can be taken multiple in a day and can be duplicates. Do we want to delete such duplicated?

Reply:
Generally, yes, we probably want to delete those types of duplicates.

Reply:
Yeah, generally we want to remove those types of duplicates. We will have information loss either way and that's the best information to drop.

Reply:
Sounds good to me. Thank you for the clarification.